backend/trustymain/app.go

```diff
@@ -0,0 +1,228 @@
+package trustymain
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/signal"
+	"runtime"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/go-phorce/dolly/netutil"
+	"github.com/go-phorce/dolly/xlog"
+	"github.com/go-phorce/dolly/xlog/logrotate"
+	"github.com/go-phorce/trusty/backend/trustyserver"
+	"github.com/go-phorce/trusty/config"
+	"github.com/go-phorce/trusty/version"
+	"github.com/juju/errors"
+	"go.uber.org/dig"
+	kp "gopkg.in/alecthomas/kingpin.v2"
+)
+
+var logger = xlog.NewPackageLogger("github.com/go-phorce/trusty/backend", "trusty")
+
+// appFlags specifies application flags
+type appFlags struct {
+	cfgFile  *string
+	cpu      *string
+	isStderr *bool
+	dryRun   *bool
+}
+
+// App provides application container
+type App struct {
+	sigs      chan os.Signal
+	container *dig.Container
+	closers   []io.Closer
+	closed    bool
+	lock      sync.RWMutex
+
+	args         []string
+	flags        *appFlags
+	cfg          *config.Configuration
+	trustyServer *trustyserver.TrustyServer
+}
+
+// New returns new App
+func New(args []string) *App {
+	return &App{
+		container: dig.New(),
+		args:      args,
+		closers:   make([]io.Closer, 0, 8),
+		flags:     new(appFlags),
+	}
+}
+
+// WithSignal adds cusom signal channel
+func (a *App) WithSignal(sigs chan os.Signal) *App {
+	a.sigs = sigs
+	return a
+}
+
+// OnClose adds a closer to be called when application exists
+func (a *App) OnClose(closer io.Closer) {
+	if closer == nil {
+		return
+	}
+
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	a.closers = append(a.closers, closer)
+}
+
+// Close implements Closer interface to clean up resources
+func (a *App) Close() error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	if a.closed {
+		return errors.Trace(trustyserver.ErrAlreadyClosed)
+	}
+
+	a.closed = true
+	// close in reverse order
+	for i := len(a.closers) - 1; i >= 0; i-- {
+		closer := a.closers[i]
+		if closer != nil {
+			err := closer.Close()
+			if err != nil {
+				logger.Errorf("src=Close, err=[%v]", err.Error())
```

Comment:
qq, if we are ignoring the error, should we ignore it everywhere (see above when a.Closed is true) and just make Close return nothing instead of error?

Reply:
i think errors in `Close` are mostly to log and analyze that the method is not called twice, etc.
I guess we should not propagate it to the caller.